src/storage/QueryBaseProcessor.inl

```diff
@@ -182,20 +97,215 @@ cpp2::ErrorCode QueryBaseProcessor<REQ, RESP>::checkAndBuildContexts(const REQ&
+            return false;
+        };
+        getters.getInputProp = [] (const std::string& prop) -> VariantType {
+            LOG(FATAL) << "Unsupport get input prop " << prop;
+            return false;
+        };
+    }
     return cpp2::ErrorCode::SUCCEEDED;
 }
 
+template<typename REQ, typename RESP>
+bool QueryBaseProcessor<REQ, RESP>::checkExp(const Expression* exp) {
+    switch (exp->kind()) {
+        case Expression::kPrimary:
+            return true;
+        case Expression::kFunctionCall:
+            // TODO(heng): we should support it in the future.
+            return false;
+        case Expression::kUnary: {
+            auto* unaExp = static_cast<const UnaryExpression*>(exp);
+            return checkExp(unaExp->operand());
+        }
+        case Expression::kTypeCasting: {
+            auto* typExp = static_cast<const TypeCastingExpression*>(exp);
+            return checkExp(typExp->operand());
+        }
+        case Expression::kArithmetic: {
+            auto* ariExp = static_cast<const ArithmeticExpression*>(exp);
+            return checkExp(ariExp->left()) && checkExp(ariExp->right());
+        }
+        case Expression::kRelational: {
+            auto* relExp = static_cast<const RelationalExpression*>(exp);
+            return checkExp(relExp->left()) && checkExp(relExp->right());
+        }
+        case Expression::kLogical: {
+            auto* logExp = static_cast<const LogicalExpression*>(exp);
+            return checkExp(logExp->left()) && checkExp(logExp->right());
+        }
+        case Expression::kSourceProp: {
+            auto* sourceExp = static_cast<const SourcePropertyExpression*>(exp);
+            const auto& tagName = sourceExp->tag();
+            const auto& propName = sourceExp->prop();
+            auto tagRet = this->schemaMan_->toTagID(spaceId_, tagName);
+            if (!tagRet.ok()) {
+                VLOG(1) << "Can't find tag " << tagName << ", in space " << spaceId_;
+                return false;
+            }
+            auto tagId = tagRet.value();
+            // TODO(heng): Now we use the latest version.
+            auto schema = this->schemaMan_->getTagSchema(spaceId_, tagId);
+            CHECK(!!schema);
+            auto field = schema->field(propName);
+            if (field == nullptr) {
+                VLOG(1) << "Can't find related prop " << propName << " on tag " << tagName;
+                return false;
+            }
+            // TODO(heng): Now we have to scan the whole array to find related tagId,
+            // maybe we could find a better way to solve it.
+            for (auto& tc : tagContexts_) {
+                if (tc.tagId_ == tagId) {
+                    auto* prop = tc.findProp(propName);
+                    if (prop == nullptr) {
+                        tc.pushFilterProp(tagName, propName, field->getType());
+                    } else if (!prop->filtered()) {
```

Comment:
the function pushFilterProp will call setTagFilterKey and the filtered is true, so maybe here no need check?